Unhardcode Explosion "Image"

reaperrr opened the pull request:
Adds `ExplosionCollection` property to `CreateEffectWarhead`.

Additionally - although I didn't test it - going by the code I believe that before my fix, `Explodes`, `Parachutable`, `WithBuildingExplosion` would only pass lint tests but NOT work in-game with anything other than "explosion", because that image string was still hardcoded on the effect itself.

Should make life easier for 3rd-party mods.

reaperrr committed:
Unhardcode Explosion "Image"

Add ExplosionCollection property to CreateEffectWarhead

Phrohdoh commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -19,7 +19,10 @@ namespace OpenRA.Mods.Common.Warheads
+		[SequenceReference("ExplosionCollection")] public readonly string[] Explosions = new string[0];
+
+		[Desc("Explosion collection containing effect.")]
+		public readonly string ExplosionCollection = "explosion";
```
Do we use `Collection` anywhere else?
This doesn't sound familiar to me at all.

reaperrr replied:
Well, Parachutable and WithBuildingExplosion already did.

The main reason is that `Image` would be somewhat unfitting given how the explosion sequences are set up.

pchote replied:
Those two traits will need fixing too.  The `Collection` term is used for UI artwork, not sequences.

reaperrr committed:
Upgrade rule for *Collection -> Image rename

reaperrr commented:
Updated.

Did away with the prefixes and went for just `Image` for the sake of consistency.
Added upgrade rule as a bonus, although that's probably unnecessary since everything else than "explosion" would currently crash on bleed, so I suspect no 3rd-party mod has tried to use custom values anyway.

Phrohdoh commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -101,8 +104,8 @@ public override void DoImpact(Target target, Actor firedBy, IEnumerable<int> dam
 			var explosion = Explosions.RandomOrDefault(Game.CosmeticRandom);
-			if (explosion != null)
-				world.AddFrameEndTask(w => w.Add(new Explosion(w, pos, explosion, palette)));
+			if (Image != null && explosion != null)
```
You'll probably want to check `string.IsNullOrEmpty` instead.

Mailaender replied:
I can't point to the line in `FieldLoader` where it is decided, but my experience is that it returns `null` instead of empty strings in all relevant cases.

Mailaender commented:
Exposing this to https://github.com/OpenRA/OpenRA/wiki/Traits documentation is always a good idea. :+1:

obrakmann commented:
Looks good to me, :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/80b652d9a135b644e88830e887f41aaf7dc60430)
